cumulus/client/cirrus-executor/src/lib.rs

```diff
@@ -388,22 +399,78 @@ where
 	/// Checks the execution receipt from the executor peers.
 	fn on_execution_receipt(
 		&self,
-		_execution_receipt: &ExecutionReceipt<<Block as BlockT>::Hash>,
+		execution_receipt: &ExecutionReceipt<<Block as BlockT>::Hash>,
 	) -> Result<Action, Self::Error> {
 		// TODO: validate the Proof-of-Election
 
-		// TODO: check if the received ER is same with the one produced locally.
-		let same_with_produced_locally = true;
-
-		if same_with_produced_locally {
-			Ok(Action::RebroadcastExecutionReceipt)
+		// TODO: more efficient execution receipt checking strategy?
+		let local_receipt = if let Some(local_receipt) =
+			crate::aux_schema::load_execution_receipt::<_, Block>(
+				&*self.client,
+				execution_receipt.secondary_hash,
+			)? {
+			local_receipt
 		} else {
+			// Wait for the local execution receipt until it's ready.
+			let (tx, rx) = crossbeam::channel::bounded::<
+				sp_blockchain::Result<ExecutionReceipt<Block::Hash>>,
+			>(1);
+			let client = self.client.clone();
+			let block_hash = execution_receipt.secondary_hash;
+			self.spawner.spawn(
+				"wait-for-local-execution-receipt",
+				None,
+				async move {
+					loop {
+						match crate::aux_schema::load_execution_receipt::<_, Block>(
+							&*client, block_hash,
+						) {
+							Ok(Some(local_receipt)) => {
+								let _ = tx.send(Ok(local_receipt));
+								break
+							},
+							Ok(None) => {
+								tokio::time::sleep(std::time::Duration::from_millis(100)).await;
+							},
+							Err(e) => {
+								let _ = tx.send(Err(e));
+								break
+							},
+						}
+					}
+				}
+				.boxed(),
+			);
+
+			rx.recv()??
+		};
+
+		// TODO: What happens for this obvious error?
+		if local_receipt.trace.len() != execution_receipt.trace.len() {}
+
+		let find_invalid_transition = || {
+			for ((local_idx, local_root), (_external_idx, external_root)) in local_receipt
+				.trace
+				.iter()
+				.enumerate()
+				.zip(execution_receipt.trace.iter().enumerate())
+			{
+				if local_root != external_root {
+					return Some(local_idx)
+				}
+			}
+			None
+		};
+
+		if let Some(_local_trace_idx) = find_invalid_transition() {
```

Comment:
I'd use `.find()` on iterator instead of wrapping into a function with `return` to make it more idiomatic

Reply:
Updated.